### PR TITLE
Implement LocalConnection

### PR DESCRIPTION
**Pull request by tschneidereit**
Not finished, contains only the content-facing parts. Those should be pretty much done, though, modulo bug fixes.

[Review on Reviewable](https://reviewable.io/reviews/mozilla/shumway/2225)

### CODE REVIEW

File: extension/firefox/chrome/ShumwayCom.jsm [hunk 1]
```diff
@@ -273,6 +291,98 @@ var ShumwayCom = {
           return;
         }
         onSystemResourceCallback = callback;
+      },
+      createLocalConnection: function(connectionName, callback) {
```
Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: Move this code into separate jsm-files (similar to RtmpUtils and SpecialInflateUtils).  You can create e.g. `localConnections` object that will encapsulate the logic, the _getLocalConnection and localConnectionsRegistry can be moved there as well.

File: extension/firefox/chrome/ShumwayCom.jsm [hunk 2]
```diff
@@ -273,6 +291,98 @@ var ShumwayCom = {
...
+        var connection = {
+          callback,
+          allowedSecureDomains: Object.create(null),
+          allowedInsecureDomains: Object.create(null)
+        };
+        localConnectionsRegistry[connectionName] = connection;
+        return 0; // LocalConnectionConnectResult.Success
+      },
+      hasLocalConnection: function(connectionName) {
+        return !!_getLocalConnection(connectionName);
```
Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: validate connectionName here and below

File: extension/firefox/chrome/ShumwayCom.jsm [hunk 3]
```diff
@@ -273,6 +291,98 @@ var ShumwayCom = {
...
+      createLocalConnection: function(connectionName, callback) {
+        if (typeof connectionName !== 'string' ||
+            connectionName[0] !== '_' && connectionName.split(':').length !== 2) {
+          return -1; // LocalConnectionConnectResult.InvalidName
+        }
+        if (localConnectionsRegistry[connectionName]) {
+          return -2; // LocalConnectionConnectResult.AlreadyTaken
+        }
+        var connection = {
+          callback,
```
Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: We did not decide if we going to use new ECMAScript syntax yet, could you change that to ES5 friendly one?

File: extension/firefox/chrome/ShumwayCom.jsm [hunk 4]
```diff
@@ -273,6 +291,98 @@ var ShumwayCom = {
...
+          return null;
+        }
+        var connection = _getLocalConnection(connectionName);
+        if (!connection) {
+          return null;
+        }
+        try {
+          var parsedURL = NetUtil.newURI(senderURL);
+          var senderDomain = parsedURL.host;
+          var allowedDomains = parsedURL.schemeIs('https') ?
```
Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: I think there is a wrong logic here: https can reach http targets.

File: test/harness/slave.js [hunk 5]
```diff
@@ -113,6 +113,8 @@ function loadMovie(path, reportFrames) {
     Shumway.SystemResourcesLoadingService.instance =
       new Shumway.Player.BrowserSystemResourcesLoadingService(builtinPath, playerglobalInfo);
 
+    Shumway.LocalConnectionService.instance = new Shumway.Player.BrowserLocalConnectionService();
```
Comment (yurydelendik) [inline test/harness/slave.js]: BrowserLocalConnectionService is not defined here

File: extension/firefox/chrome/ShumwayCom.jsm [hunk 6]
```diff
@@ -273,6 +291,98 @@ var ShumwayCom = {
...
+          var callback = connection.callback;
+          var clonedArgs = Components.utils.cloneInto(argsBuffer, callback);
+          return Components.utils.cloneInto(callback(methodName, clonedArgs), sender);
+        } catch (e) {
+          return Components.utils.cloneInto({name: 'Error', $Bgmessage: 'Unknown error occurred',
+                                             _errorID: 0}, sender);
+        }
+      },
+      allowDomainsForLocalConnection: function(connectionName, domains, secure) {
+        var connection = _getLocalConnection(connectionName);
```
Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: Validate domains and secure.

File: extension/firefox/chrome/ShumwayCom.jsm [hunk 7]
```diff
@@ -273,6 +291,98 @@ var ShumwayCom = {
...
+      },
+      closeLocalConnection: function(connectionName) {
+        if (!_getLocalConnection(connectionName)) {
+          return -1; // LocalConnectionCloseResult.NotConnected
+        }
+        delete localConnectionsRegistry[connectionName];
+        return 0; // LocalConnectionCloseResult.Success
+      },
+      sendLocalConnectionMessage: function(connectionName, methodName, argsBuffer, sender,
+                                           senderURL) {
```
Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: We shall not trust the senderURL, it shall be taken from object swf url for now (until we figure out how connect security domains and chrome code)

File: extension/firefox/chrome/ShumwayCom.jsm (hunk 1 above)
Comment (tschneidereit) [inline extension/firefox/chrome/ShumwayCom.jsm]: Yes, I had planned on doing that.

File: extension/firefox/chrome/ShumwayCom.jsm (hunk 2 above)
Comment (tschneidereit) [inline extension/firefox/chrome/ShumwayCom.jsm]: Thanks for pointing out the missing validations. I'm aware of at least some of them and intend to move all validation into dedicated functions before landing.

File: extension/firefox/chrome/ShumwayCom.jsm (hunk 3 above)
Comment (tschneidereit) [inline extension/firefox/chrome/ShumwayCom.jsm]: Good point. Is there any reason not to use these features, though? They've been available a few versions back already, and all of this code is Firefox-only anyway? We should probably make explicit decisions on which features to use, but I do think that some are just too nice to pass on them.

File: extension/firefox/chrome/ShumwayCom.jsm (hunk 4 above)
Comment (tschneidereit) [inline extension/firefox/chrome/ShumwayCom.jsm]: ooh, I see what you mean. So insecureDomains should always be a perfect subset of secureDomains? That'd be easy enough to do at the point where the domains are added by just adding the secure domains to both sets. It's not possible to revoke access, so we don't have to worry about removal from the sets.

File: test/harness/slave.js (hunk 5 above)
Comment (tschneidereit) [inline test/harness/slave.js]: Thanks, noted. Left-over cruft from earlier versions.

File: extension/firefox/chrome/ShumwayCom.jsm (hunk 7 above)
Comment (tschneidereit) [inline extension/firefox/chrome/ShumwayCom.jsm]: Right, here's my idea for that: when we load a SWF into a player instance, we note the URL In `sendLocalConnectionMessage`, we check if `senderURL` is in the set of remembered URLs for that player instance. If not, we reject here. Do you think that makes sense?

Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: I don't think "remembered URLs" concept is clearly understood yet. But yes, you can compare/validate senderURL with SWF URL atm.

Comment (tschneidereit) [inline extension/firefox/chrome/ShumwayCom.jsm]: > I don't think "remembered URLs" concept is clearly understood yet. 

I'm not sure I understand what you mean by that. We don't current record the URLs of things we load, but that wouldn't be hard to change.

> But yes, you can compare/validate senderURL with SWF URL atm.

I do agree that we should start out with that and will implement it. Everything else can come later.

Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: > We don't current record the URLs of things we load, but that wouldn't be hard to change.

Loading does not mean it has rights to send messages, e.g. swf loads another domain swf that opens a localconnection. Other swfs from first swf's domain shall not be able to send messages to the latter, no?

Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: ... which also might be a reason to not open localconnections for loaded swfs :/

File: extension/firefox/chrome/ShumwayCom.jsm (hunk 1 above)
Comment (yurydelendik) [inline extension/firefox/chrome/ShumwayCom.jsm]: Also, place it under the preference for now.

File: extension/firefox/chrome/ShumwayCom.jsm (hunk 7 above)
Comment (tschneidereit) [inline extension/firefox/chrome/ShumwayCom.jsm]: > Loading does not mean it has rights to send messages, e.g. swf loads another domain swf that opens a localconnection. Other swfs from first swf's domain shall not be able to send messages to the latter, no?

Can you describe the attack scenario you have in mind here? I see the point, but can't come up with a scenario where this actualy matters.

**Comment by yurydelendik** [comment]
Review status: 11 of 36 files reviewed, 6 unresolved discussions, some commit checks failed.
Reviewed files:
- examples/inspector/js/inspectorPlayer.js @ r3
- extension/firefox/content/web/viewerPlayer.js @ r2
- src/avm2/errors.ts @ r2
- src/flash/events/AsyncErrorEvent.as @ r2
- src/flash/events/AsyncErrorEvent.ts @ r2
- src/flash/events/StatusEvent.as @ r2
- src/flash/events/StatusEvent.ts @ r2
- src/flash/link.ts @ r2
- src/flash/module.ts @ r2
- src/flash/net/LocalConnection.as @ r2
- src/flash/references.ts @ r2

---

<sup>**[extension/firefox/chrome/ShumwayCom.jsm, line 120 \[r3\]](https://reviewable.io:443/reviews/mozilla/shumway/2225#-JqVmyoLHmjdepcet4JW)** ([raw file](https://github.com/mozilla/shumway/blob/8b5be8bd2803ea9550dcb1911ff1dde1685ccfcb/extension/firefox/chrome/ShumwayCom.jsm#L120)):</sup>
Unneeded change

---

<sup>**[src/base/external.ts, line 54 \[r3\]](https://reviewable.io:443/reviews/mozilla/shumway/2225#-JqVnWtbYJZ3nH9knV_p)** ([raw file](https://github.com/mozilla/shumway/blob/8b5be8bd2803ea9550dcb1911ff1dde1685ccfcb/src/base/external.ts#L54)):</sup>
replace all LocalConnection members with `getLocalConnectionManager` which will return an object (as example see createSpecialStorage above) with these methods.

---

<sup>**[src/flash/display/LoaderInfo.ts, line 129 \[r3\]](https://reviewable.io:443/reviews/mozilla/shumway/2225#-JqVoEFKBHZoSjmGs-C5)** ([raw file](https://github.com/mozilla/shumway/blob/8b5be8bd2803ea9550dcb1911ff1dde1685ccfcb/src/flash/display/LoaderInfo.ts#L129)):</sup>
We shall do it before constructing LoaderInfo.

---

<sup>**[src/flash/references.ts, line 64 \[r3\]](https://reviewable.io:443/reviews/mozilla/shumway/2225#-JqVpE7Xw1t-aajQ1xyn)** ([raw file](https://github.com/mozilla/shumway/blob/8b5be8bd2803ea9550dcb1911ff1dde1685ccfcb/src/flash/references.ts#L64)):</sup>
nit: remove extra space before '///'

---

<sup>**[src/flash/references.ts, line 91 \[r3\]](https://reviewable.io:443/reviews/mozilla/shumway/2225#-JqVpMpx7dIIjS3PVJYY)** ([raw file](https://github.com/mozilla/shumway/blob/8b5be8bd2803ea9550dcb1911ff1dde1685ccfcb/src/flash/references.ts#L91)):</sup>
nit: remove extra space before '///'

---

<sup>**[src/player/player.ts, line 61 \[r3\]](https://reviewable.io:443/reviews/mozilla/shumway/2225#-JqVpX-2Q59yvX81q0y6)** ([raw file](https://github.com/mozilla/shumway/blob/8b5be8bd2803ea9550dcb1911ff1dde1685ccfcb/src/player/player.ts#L61)):</sup>
This entire class move is unrelated to LocalConnection, so skipping review for that. Please submit separate PR if review of that part is needed.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/shumway/2225)

**Comment by yurydelendik** [comment]
Okay, reviewable.io does not allow review this PR 

screen shot 2015-05-29 at 1 06 15 pm

Could you submit a PR that will just move Player class? It's not related to this PR. If this PR depends on this, let's block it by the Player class refactoring.

(Also, please notice that Travis is failing)

**Comment by tschneidereit** [comment]
Closing because reviewable.io really doesn't like this PR. Probably caused by first renaming player.ts and then doing the renaming in another PR, rebasing this one on top of that. Ideally, reviewable should just ignore the entire renaming operation, but it doesn't. Will open a fresh PR.

**Pull request closed without merging**